File: web/lib/greeks.ts
// Griegas Black-Scholes para el proveedor de Time & Sales por Databento.
//
// Por qué: OPRA (Databento) entrega solo trade + BBO, SIN delta ni IV. MarketSnack sí los
// daba ya calculados. Para producir el mismo `RawTrade` sin depender de MarketSnack, aquí
// calculamos delta/IV/gamma/theta/vega a partir del precio REAL de la operación y el precio
// del subyacente en ese instante.
//
// Convención alineada con el resto del proyecto de Victor: r = 0, sin dividendos
// (igual que `bsGamma` en gex.ts), para que las griegas sean consistentes con el GEX.

import { normCdf } from "./expectedMove";
import { bsGamma } from "./gex";

export { bsGamma };

/** Densidad normal estándar φ(x). */
function phi(x: number): number {
  return Math.exp(-0.5 * x * x) / Math.sqrt(2 * Math.PI);
}

function d1of(spot: number, strike: number, T: number, iv: number): number {
  return (Math.log(spot / strike) + 0.5 * iv * iv * T) / (iv * Math.sqrt(T));
}

function valid(spot: number, strike: number, T: number, iv: number): boolean {
  return spot > 0 && strike > 0 && T > 0 && iv > 0;
}

/** Precio Black-Scholes (r = 0, sin dividendos). En el límite devuelve el valor intrínseco. */
export function bsPrice(
  spot: number,
  strike: number,
  T: number,
  iv: number,
  isCall: boolean,
): number {
  if (!valid(spot, strike, T, iv)) {
    return Math.max(isCall ? spot - strike : strike - spot, 0);
  }
  const d1 = d1of(spot, strike, T, iv);
  const d2 = d1 - iv * Math.sqrt(T);
  return isCall
    ? spot * normCdf(d1) - strike * normCdf(d2)
    : strike * normCdf(-d2) - spot * normCdf(-d1);
}

/** Delta Black-Scholes (r = 0). null si los insumos no son válidos. */
export function bsDelta(
  spot: number,
  strike: number,
  T: number,
  iv: number,
  isCall: boolean,
): number | null {
  if (!valid(spot, strike, T, iv)) return null;
  const d1 = d1of(spot, strike, T, iv);
  return isCall ? normCdf(d1) : normCdf(d1) - 1;
}

/** Vega Black-Scholes (por 1.00 de vol, no por 1%). */
export function bsVega(spot: number, strike: number, T: number, iv: number): number {
  if (!valid(spot, strike, T, iv)) return 0;
  const d1 = d1of(spot, strike, T, iv);
  return spot * phi(d1) * Math.sqrt(T);
}

/** Theta Black-Scholes por AÑO (r = 0). Dividir entre 365 para theta diaria. */
export function bsTheta(
  spot: number,
  strike: number,
  T: number,
  iv: number,
  _isCall: boolean,
): number {
  if (!valid(spot, strike, T, iv)) return 0;
  const d1 = d1of(spot, strike, T, iv);
  // Con r = 0 el término −rK·N(d2) desaparece → theta es igual para call y put.
  return -(spot * phi(d1) * iv) / (2 * Math.sqrt(T));
}

/**
 * IV implícita por bisección a partir del precio real de la operación (r = 0).
 * Devuelve null si el precio está por debajo del intrínseco o fuera del rango alcanzable.
 */
export function impliedVol(
  price: number,
  spot: number,
  strike: number,
  T: number,
  isCall: boolean,
  lo = 1e-4,
  hi = 5,
  tol = 1e-6,
): number | null {
  if (!(price > 0) || !(spot > 0) || !(strike > 0) || !(T > 0)) return null;
  const intrinsic = Math.max(isCall ? spot - strike : strike - spot, 0);
  if (price < intrinsic - 1e-6) return null;
  if (price > bsPrice(spot, strike, T, hi, isCall)) return null;
  let a = lo;
  let b = hi;
  for (let i = 0; i < 100; i++) {
    const mid = 0.5 * (a + b);
    const p = bsPrice(spot, strike, T, mid, isCall);
    if (Math.abs(p - price) < tol) return mid;
    if (p > price) b = mid;
    else a = mid;
  }
  return 0.5 * (a + b);
}

/** Todas las griegas de una operación, calculadas desde su precio real y el subyacente. */
export interface TradeGreeks {
  iv: number | null;
  delta: number | null;
  gamma: number;
  theta: number; // por año
  vega: number;
}

export function tradeGreeks(
  tradePrice: number,
  spot: number,
  strike: number,
  T: number,
  isCall: boolean,
): TradeGreeks {
  const iv = impliedVol(tradePrice, spot, strike, T, isCall);
  if (iv == null) return { iv: null, delta: null, gamma: 0, theta: 0, vega: 0 };
  return {
    iv,
    delta: bsDelta(spot, strike, T, iv, isCall),
    gamma: bsGamma(spot, strike, T, iv),
    theta: bsTheta(spot, strike, T, iv, isCall),
    vega: bsVega(spot, strike, T, iv),
  };
}

File: webapp/backend/greeks.py
"""Black-Scholes: IV implícita y delta EN EL MOMENTO DE CADA OPERACIÓN.

Por qué así: la delta no es un dato observable del mercado, es una cantidad de modelo.
La forma correcta de obtener "la delta de ese trade" es:
  1) tomar el precio REAL al que se operó (del tape exacto de Databento),
  2) el precio REAL del subyacente en ese minuto (Massive),
  3) la tasa libre de riesgo de ese día (FRED),
  4) despejar la IV implícita de ese precio, y con ella calcular la delta.

Así la delta corresponde al instante del trade, no a un snapshot de hoy.

Supuestos declarados: Black-Scholes europeo, sin rendimiento por dividendos (q=0 por
defecto, configurable). Para opciones de corto plazo el efecto es mínimo.
"""
from __future__ import annotations

import math

SQRT2 = math.sqrt(2.0)


def _norm_cdf(x: float) -> float:
    return 0.5 * (1.0 + math.erf(x / SQRT2))


def _d1(S: float, K: float, T: float, r: float, sigma: float, q: float = 0.0) -> float:
    return (math.log(S / K) + (r - q + 0.5 * sigma * sigma) * T) / (sigma * math.sqrt(T))


def bs_price(S: float, K: float, T: float, r: float, sigma: float,
             is_call: bool, q: float = 0.0) -> float:
    if T <= 0 or sigma <= 0 or S <= 0 or K <= 0:
        # valor intrínseco en el límite
        intrinsic = (S - K) if is_call else (K - S)
        return max(intrinsic, 0.0)
    d1 = _d1(S, K, T, r, sigma, q)
    d2 = d1 - sigma * math.sqrt(T)
    if is_call:
        return S * math.exp(-q * T) * _norm_cdf(d1) - K * math.exp(-r * T) * _norm_cdf(d2)
    return K * math.exp(-r * T) * _norm_cdf(-d2) - S * math.exp(-q * T) * _norm_cdf(-d1)


def bs_delta(S: float, K: float, T: float, r: float, sigma: float,
             is_call: bool, q: float = 0.0) -> float | None:
    if T <= 0 or sigma <= 0 or S <= 0 or K <= 0:
        return None
    d1 = _d1(S, K, T, r, sigma, q)
    disc = math.exp(-q * T)
    return disc * _norm_cdf(d1) if is_call else disc * (_norm_cdf(d1) - 1.0)


def implied_vol(price: float, S: float, K: float, T: float, r: float,
                is_call: bool, q: float = 0.0,
                lo: float = 1e-4, hi: float = 5.0, tol: float = 1e-6) -> float | None:
    """Despeja la IV por bisección a partir del precio real de la operación."""
    if price is None or price <= 0 or T <= 0 or S <= 0 or K <= 0:
        return None
    intrinsic = max((S - K) if is_call else (K - S), 0.0)
    if price < intrinsic - 1e-6:      # precio por debajo del intrínseco -> sin solución
        return None
    p_hi = bs_price(S, K, T, r, hi, is_call, q)
    if price > p_hi:                   # fuera del rango de vol alcanzable
        return None
    for _ in range(100):
        mid = 0.5 * (lo + hi)
        p = bs_price(S, K, T, r, mid, is_call, q)
        if abs(p - price) < tol:
            return mid
        if p > price:
            hi = mid
        else:
            lo = mid
    return 0.5 * (lo + hi)
